src/deckgl-layers/layer-utils/cpu-aggregator.js

```diff
@@ -0,0 +1,606 @@
+  } = step;
+
+  if (!this.state.dimensions[key].sortedBins) {
+    // the previous step should set sortedBins, if not, something went wrong
+    return;
+  }
+
+  // for log and sqrt scale, returns linear domain by default
+  // TODO: support other scale function domain in bin sorter
+  const valueDomain = this.state.dimensions[
+    key
+  ].sortedBins.getValueDomainByScale(props[scaleType.prop], [
+    props[lowerPercentile.prop],
+    props[upperPercentile.prop]
+  ]);
+
+  this._setDimensionState(key, {valueDomain});
+}
+
+export function getDimensionScale(step, props, dimensionUpdater) {
+  const {key} = dimensionUpdater;
+  const {domain, range, scaleType} = step.triggers;
+  const {onSet} = step;
+  if (!this.state.dimensions[key].valueDomain) {
+    // the previous step should set valueDomain, if not, something went wrong
+    return;
+  }
+
+  const dimensionRange = props[range.prop];
+  const dimensionDomain =
+    props[domain.prop] || this.state.dimensions[key].valueDomain;
+
+  const scaleFunctor = getScaleFunctor(
+    scaleType && props[scaleType.prop]
+  )();
+
+  const scaleFunc = scaleFunctor
+    .domain(dimensionDomain)
+    .range(dimensionRange);
+
+  if (typeof onSet === 'object' && typeof props[onSet.props] === 'function') {
+    props[onSet.props](scaleFunc.domain());
+  }
+  this._setDimensionState(key, {scaleFunc});
+}
+
+function normalizeResult(result = {}) {
+  // support previous hexagonAggregator API
+  if (result.hexagons) {
+    return Object.assign({data: result.hexagons}, result);
+  } else if (result.layerData) {
+    return Object.assign({data: result.layerData}, result);
+  }
+
+  return result;
+}
+
+export function getAggregatedData(step, props, aggregation, viewport) {
+  const {
+    triggers: {aggregator: aggr}
+  } = step;
+  const aggregator = props[aggr.prop];
+
+  // result should contain a data array and other props
+  // result = {data: [], ...other props}
+  const result = aggregator(props, viewport);
+  this.setState({
+    layerData: normalizeResult(result)
+  });
+}
+
+export const defaultAggregation = {
+  key: 'position',
+  updateSteps: [
+    {
+      key: 'aggregate',
+      triggers: {
+        cellSize: {
+          prop: 'cellSize'
+        },
+        position: {
+          prop: 'getPosition',
+          updateTrigger: 'getPosition'
+        },
+        aggregator: {
+          prop: 'gridAggregator'
+        }
+      },
+      updater: getAggregatedData
+    }
+  ]
+};
+
+function getSubLayerAccessor(dimensionState, dimension, layerProps) {
+  return cell => {
+    const {sortedBins, scaleFunc} = dimensionState;
+    const bin = sortedBins.binMap[cell.index];
+
+    if (bin && bin.counts === 0) {
+      // no points left in bin after filtering
+      return dimension.nullValue;
+    }
+
+    const cv = bin && bin.value;
+    const domain = scaleFunc.domain();
+
+    const isValueInDomain = cv >= domain[0] && cv <= domain[domain.length - 1];
```

Comment:
i think we perform value in domain checks in several places, would it be better to create an helper?

Reply:
this is in deck.gl